lib/rubygems/sigstore/openid/dynamic.rb

```diff
@@ -0,0 +1,208 @@
+      server = TCPServer.new 5678
+      server_addr = "5678"
+    else
+      server = TCPServer.new 0
+      server_addr = server.addr[1].to_s
+    end
+
+    webserv = Thread.new do
+      begin
+        response = "You may close this browser"
+        response_code = "200 OK"
+        connection = server.accept
+        while (input = connection.gets)
+          begin
+            # VERB PATH HTTP/1.1
+            http_req = input.split(' ')
+            if http_req.length != 3
+              raise "invalid HTTP request received on callback"
+            end
+            params = CGI.parse(URI.parse(http_req[1]).query)
+            if params["code"].length != 1 or params["state"].length != 1
+              raise "multiple values for code or state returned in callback; unable to process"
+            end
+            Thread.current[:code] = params["code"][0]
+            Thread.current[:state] = params["state"][0]
+          rescue StandardError => e
+            response = "Error processing request: #{e.message}"
+            response_code = "400 Bad Request"
+          end
+          connection.print "HTTP/1.1 #{response_code}\r\n" +
+                      "Content-Type: text/plain\r\n" +
+                      "Content-Length: #{response.bytesize}\r\n" +
+                      "Connection: close\r\n"
+          connection.print "\r\n"
+          connection.print response
+          connection.close
+          if response_code != "200 OK"
+            raise response
+          end
+          break
+        end
+      ensure
+        server.close
+      end
+    end
+
+    webserv.abort_on_exception = true
+    redirect_uri = "http://localhost:" + server_addr
+
+    client = OpenIDConnect::Client.new(
+      authorization_endpoint: oidc_discovery.authorization_endpoint,
+      identifier: config.oidc_client,
+      redirect_uri: redirect_uri,
+      secret: config.oidc_secret,
+      token_endpoint: oidc_discovery.token_endpoint,
+    )
+
+    authorization_uri = client.authorization_uri(
+      scope: ["openid", :email],
+      state: session[:state],
+      nonce: session[:nonce],
+      code_challenge_method: pkce[:method],
+      code_challenge: pkce[:challenge],
+    )
+
+    begin
+      if ENV["SIGSTORE_TEST"]
+        Faraday.get("#{redirect_uri}/?code=DUMMY&state=#{session[:state]}")
+      else
+        Launchy.open(authorization_uri)
+      end
+    rescue
+      # NOTE: ignore any exception, as the URL is printed above and may be
+      #       opened manually
+      say "Cannot open browser automatically, please click on the link below:"
+      say ""
+      say authorization_uri
+    end
+
+    webserv.join
+
+    # check state == webserv[:state]
+    if webserv[:state] != session[:state]
+      abort 'Invalid state value received from OIDC Provider'
+    end
+
+    client.authorization_code = webserv[:code]
+    @access_token = client.access_token!({code_verifier: pkce[:value]})
```

Comment:
I think we should only assign the instance variable once we have returned from `verify_token`.  Otherwise, the OpenId::Dynamic instance may still serve up a bad access token.